src/main/java/ca/on/oicr/pineryreports/reports/impl/LibrariesBillingReport.java

```diff
@@ -223,24 +225,47 @@ private DetailedObject makeDetailedRow(SampleDto lib, Map<String, SampleDto> all
         (lib.getPreparationKit() == null ? "No Kit" : lib.getPreparationKit().getName()),
         lib.getCreatedDate(),
         lib.getName(),
-        getAttribute(LIBRARY_DESIGN_CODE, lib));
+        getAttribute(LIBRARY_DESIGN_CODE, lib),
+        getExternalNames(lib, allSamples));
   }
 
+  private String getExternalNames(SampleDto lib, Map<String, SampleDto> allSamples) {
+    return SampleUtils.getUpstreamAttribute(ATTR_EXTERNAL_NAME, lib, allSamples);
+  }
+
+  /**
+   * This method is used both to output the first header row of the report, and for calculating the
+   * number of columns in order to ensure a rectangular CSV. Since our two tables in this report are
+   * of differing widths, find the wider of the two and append blanks for the correct width.
+   */
   @Override
   protected List<ColumnDefinition> getColumns() {
+    List<ColumnDefinition> columnDefinitions = new LinkedList<>();
+    List<String> summaryHeadings = getSummaryHeadings();
+    int largestHeadingsSize = Math.max(summaryHeadings.size(), getDetailedHeadings().size());
+
+    for (int i = 0; i < largestHeadingsSize; i++) {
+      try {
+        columnDefinitions.add(new ColumnDefinition(summaryHeadings.get(i)));
+      } catch (IndexOutOfBoundsException ioob) {
```

Comment:
Should use `if (i < summaryHeadings.size())` instead of intentionally throwing and catching an exception to control normal/expected operation

Reply:
done